src/filesystem/implementations/as.h

```diff
@@ -175,72 +177,79 @@ ASFileSystem::CheckClient()
-                                    path + ", errno:" + strerror(errno));
+        Status::Code::INTERNAL,
+        "Unable to get blob property for file at " + path + ":" + ex.what());
   }
 
-  auto time =
-      std::chrono::system_clock::from_time_t(blobProperty.last_modified);
-  auto update_time =
-      std::chrono::time_point_cast<std::chrono::nanoseconds>(time)
-          .time_since_epoch()
-          .count();
-
-  *mtime_ns = update_time;
   return Status::Success;
 };
 
 Status
 ASFileSystem::ListDirectory(
     const std::string& container, const std::string& dir_path,
-    std::function<
-        Status(const as::list_blobs_segmented_item&, const std::string&)>
-        func)
+    std::function<Status(
+        const std::vector<asb::Models::BlobItem>& blobs,
+        const std::vector<std::string>& blob_prefixes)>
+        callback)
 {
-  as::blob_client_wrapper bc(client_);
-
+  auto container_client = client_->GetBlobContainerClient(container);
+  auto options = asb::ListBlobsOptions();
   // Append a slash to make it easier to list contents
   std::string full_dir = AppendSlash(dir_path);
-  auto blobs = bc.list_blobs_segmented(container, "/", "", full_dir);
-  if (errno != 0) {
-    return Status(
-        Status::Code::INTERNAL, "Failed to get contents of directory " +
-                                    dir_path + ", errno:" + strerror(errno));
-  }
+  options.Prefix = full_dir;
 
-  for (auto&& item : blobs.blobs) {
-    std::string name = item.name;
-    int item_start = name.find(full_dir) + full_dir.size();
-    int item_end = name.find("/", item_start);
-    // Let set take care of subdirectory contents
-    std::string subfile = name.substr(item_start, item_end - item_start);
-    auto status = func(item, subfile);
-    if (!status.IsOk()) {
-      return status;
+  try {
+    for (auto blobPage = container_client.ListBlobsByHierarchy("/", options);
+         blobPage.HasPage(); blobPage.MoveToNextPage()) {
+      // per-page per-blob
+      RETURN_IF_ERROR(callback(blobPage.Blobs, blobPage.BlobPrefixes));
     }
   }
+  catch (as::StorageException& ex) {
+    return Status(
+        Status::Code::INTERNAL,
+        "Failed to get contents of directory " + dir_path + ":" + ex.what());
+  }
+
   return Status::Success;
 }
 
 Status
 ASFileSystem::GetDirectoryContents(
     const std::string& path, std::set<std::string>* contents)
 {
-  auto func = [&](const as::list_blobs_segmented_item& item,
-                  const std::string& dir) {
-    contents->insert(dir);
-    // Fail-safe check to ensure the item name is not empty
-    if (dir.empty()) {
-      return Status(
-          Status::Code::INTERNAL,
-          "Cannot handle item with empty name at " + path);
+  auto func = [&](const std::vector<asb::Models::BlobItem>& blobs,
+                  const std::vector<std::string>& blob_prefixes) {
+    for (const auto& blob_item : blobs) {
+      // Fail-safe check to ensure the item name is not empty
```

Comment:
While these comments explaining the fail-safe check are nice, I don't feel they're necessary. Your code (error-check followed by specific error) is clear and self-explanatory.

Reply:
@kthui ^^^

Reply:
> I don't feel they're necessary.

There were GitHub issues (#5533 and #5501) where the user requested for a path e.g. `az://bucket/model_repo/model_dir/config.pbtxt`, which the `IsDirectory()` likely returned `True` for the `config.pbtxt`. Then, the code moved on to expanding the `.../config.pbtxt` path, which will resulted in an item (file or directory) that has an empty name after the `.../config.pbtxt`, because `.../config.pbtxt` is a file with no child contents afterward. This check will likely catch that if it would to happen again in the future.